In [1]:
import numpy as np
import pandas as pd

data_version = 1
df = pd.read_csv(fr"E:\datasets\eeg\data{data_version}.csv")
df.head()

Timestamp  Channel 1  Channel 2  Channel 3  Channel 4  Channel 5  \
0     0.4883  -0.008051   0.002599  -0.004225  -0.011760  -0.008265   
1     0.9766  -0.008049   0.002599  -0.004232  -0.011764  -0.008263   
2     1.4649  -0.008054   0.002596  -0.004242  -0.011769  -0.008265   
3     1.9532  -0.008062   0.002595  -0.004250  -0.011769  -0.008268   
4     2.4415  -0.008065   0.002592  -0.004256  -0.011767  -0.008268   

   Channel 6  Channel 7  Channel 8  Channel 9  ...  Channel 22  Channel 23  \
0  -0.002259  -0.007579  -0.013511  -0.004944  ...   -0.005836   -0.001075   
1  -0.002256  -0.007580  -0.013510  -0.004939  ...   -0.005840   -0.001076   
2  -0.002255  -0.007577  -0.013510  -0.004938  ...   -0.005843   -0.001072   
3  -0.002263  -0.007577  -0.013509  -0.004945  ...   -0.005847   -0.001067   
4  -0.002263  -0.007580  -0.013504  -0.004944  ...   -0.005846   -0.001069   

   Channel 24  Channel 25  Channel 26  Channel 27  Marker tag  \
0    0.008498           0      137285           0    [Marker]   
1    0.008498           0      137286           0         NaN   
2    0.008502           0      137287           0         NaN   
3    0.008498           0      137288           0         NaN   
4    0.008496           0      137289           0         NaN   

   Marker timestamp  Marker value  Unnamed: 31  
0              0.04  Palm Forward          NaN  
1               NaN           NaN          NaN  
2               NaN           NaN          NaN  
3               NaN           NaN          NaN  
4               NaN           NaN          NaN  

[5 rows x 32 columns]

In [2]:
df_notna = df[df["Marker value"].notna()]
df_notna

Timestamp  Channel 1  Channel 2  Channel 3  Channel 4  Channel 5  \
0        4.883000e-01  -0.008051   0.002599  -0.004225  -0.011760  -0.008265   
6176     3.016229e+03  -0.007984   0.002679  -0.004164  -0.011700  -0.008158   
12388    6.049549e+03  -0.008032   0.002641  -0.004169  -0.011671  -0.008186   
18566    9.066266e+03  -0.008026   0.002658  -0.004185  -0.011689  -0.008204   
24777    1.209910e+04  -0.008082   0.002585  -0.004261  -0.011722  -0.008236   
...               ...        ...        ...        ...        ...        ...   
2195831  1.072225e+06  -0.007644   0.002628  -0.004182  -0.010737  -0.008759   
2202009  1.075241e+06  -0.007595   0.002584  -0.003990  -0.010594  -0.008725   
2208186  1.078258e+06  -0.007593   0.002679  -0.004128  -0.010667  -0.008731   
2214466  1.081324e+06  -0.007575   0.002608  -0.003936  -0.010559  -0.008710   
2220677  1.084357e+06  -0.007576   0.002682  -0.004092  -0.010641  -0.008695   

         Channel 6  Channel 7  Channel 8  Channel 9  ...  Channel 22  \
0        -0.002259  -0.007579  -0.013511  -0.004944  ...   -0.005836   
6176     -0.002254  -0.007569  -0.013457  -0.004888  ...   -0.005805   
12388    -0.002246  -0.007558  -0.013407  -0.004872  ...   -0.005762   
18566    -0.002205  -0.007534  -0.013412  -0.004917  ...   -0.005813   
24777    -0.002256  -0.007579  -0.013467  -0.005031  ...   -0.005866   
...            ...        ...        ...        ...  ...         ...   
2195831  -0.002423  -0.007851  -0.012959  -0.004852  ...   -0.005281   
2202009  -0.002422  -0.007857  -0.013093  -0.005017  ...   -0.005281   
2208186  -0.002386  -0.007809  -0.012913  -0.004800  ...   -0.005277   
2214466  -0.002378  -0.007847  -0.013043  -0.004978  ...   -0.005316   
2220677  -0.002340  -0.007801  -0.012943  -0.004805  ...   -0.005294   

         Channel 23  Channel 24  Channel 25  Channel 26  Channel 27  \
0         -0.001075    0.008498           0      137285           0   
6176      -0.001052    0.008516           0      143444           0   
12388     -0.001065    0.008543           0      149635           0   
18566     -0.001105    0.008549           0      155832           0   
24777     -0.001130    0.008487           0      161994           0   
...             ...         ...         ...         ...         ...   
2195831   -0.000514    0.008098           0     2333133           0   
2202009   -0.000496    0.008109           0     2339275           0   
2208186   -0.000493    0.008126           0     2345454           0   
2214466   -0.000507    0.008114           0     2351757           0   
2220677   -0.000493    0.008123           0     2357947           0   

         Marker tag  Marker timestamp  Marker value  Unnamed: 31  
0          [Marker]              0.04  Palm Forward          NaN  
6176       [Marker]           3016.43         Pause          NaN  
12388      [Marker]           6049.66     Extension          NaN  
18566      [Marker]           9066.12         Pause          NaN  
24777      [Marker]          12099.27  Palm Forward          NaN  
...             ...               ...           ...          ...  
2195831    [Marker]        1072224.52  Palm Forward          NaN  
2202009    [Marker]        1075241.27         Pause          NaN  
2208186    [Marker]        1078257.54     Extension          NaN  
2214466    [Marker]        1081324.10         Pause          NaN  
2220677    [Marker]        1084357.27  Palm Forward          NaN  

[357 rows x 32 columns]

In [3]:
df_length = len(df)
sequence_length = (
    df_notna.index.to_series().diff().shift(-1).fillna(0).astype(int)
)
sequence_length.iloc[-1] = df_length - sequence_length.index[-1].astype(int)
df_sequences_raw = pd.DataFrame({
    "Index start": df_notna.index.astype(int),
    "Sequence length": sequence_length.values,
    "Marker value": df_notna["Marker value"].values
})
df_sequences_raw

Index start  Sequence length  Marker value
0              0             6176  Palm Forward
1           6176             6212         Pause
2          12388             6178     Extension
3          18566             6211         Pause
4          24777             6212  Palm Forward
..           ...              ...           ...
352      2195831             6178  Palm Forward
353      2202009             6177         Pause
354      2208186             6280     Extension
355      2214466             6211         Pause
356      2220677             3041  Palm Forward

[357 rows x 3 columns]

In [4]:
df_sequences_raw["Sequence length"].mean()

6228.901960784314

## Сэмплинг данных

In [5]:
batch_size = int(2048 * 1.6)  # Возьмем по 1.6 секунды (0 == вариабельные посл-сти)
start_offset = int(2048 * 0.3)  # Откинем 0.3 стартовые секунды
end_offset = int(2048 * 0.3)  # Откинем 0.3 конечные секунды
important_first = True  # Обозначим, какая часть нужнее - начало или конец, в том случае, если посл-сть меньше суммы значений сверху

df_sampled = []
for idx, row in df_sequences_raw.iterrows():
    seq_len = row['Sequence length']
    idx_start = row['Index start']
    idx_end = idx_start + seq_len
    marker_value = row['Marker value']
    
    if seq_len > batch_size > 0:
        if seq_len >= batch_size + start_offset + end_offset:
            if important_first:
                idx_start = idx_start + start_offset
                idx_end = idx_start + batch_size
            else:
                idx_end = idx_end - end_offset
                idx_start = idx_end - batch_size
        else:
            overflow = batch_size + start_offset + end_offset - seq_len
            if important_first:
                idx_start = idx_start + max(0, start_offset - overflow)
                idx_end = min(idx_start + batch_size, idx_end)
            else:
                idx_end = idx_end - max(0, end_offset - overflow)
                idx_start = max(idx_end - batch_size, idx_start)
    
    df_cur = df.iloc[idx_start:idx_end].drop(columns=[
        'Timestamp', 'Marker tag', 'Unnamed: 31',
        'Channel 25', 'Channel 26', 'Channel 27', 'Marker timestamp'
    ]).copy()
    df_cur['Sequence timestamp'] = np.arange(idx_start, idx_end, step=1, dtype='int32')
    df_cur['Marker value'] = marker_value
    df_sampled.append(df_cur)

df_sampled = pd.concat(df_sampled)
df_sampled.reset_index(drop=True, inplace=True)
df_sampled

Channel 1  Channel 2  Channel 3  Channel 4  Channel 5  Channel 6  \
0        -0.008091   0.002559  -0.004241  -0.011768  -0.008257  -0.002263   
1        -0.008086   0.002557  -0.004234  -0.011751  -0.008254  -0.002262   
2        -0.008083   0.002559  -0.004229  -0.011747  -0.008258  -0.002262   
3        -0.008082   0.002556  -0.004228  -0.011753  -0.008256  -0.002265   
4        -0.008082   0.002561  -0.004219  -0.011747  -0.008253  -0.002265   
...            ...        ...        ...        ...        ...        ...   
1169292  -0.007588   0.002673  -0.004112  -0.010642  -0.008710  -0.002375   
1169293  -0.007590   0.002673  -0.004110  -0.010642  -0.008711  -0.002378   
1169294  -0.007586   0.002673  -0.004101  -0.010644  -0.008711  -0.002377   
1169295  -0.007584   0.002671  -0.004103  -0.010641  -0.008709  -0.002375   
1169296  -0.007581   0.002674  -0.004099  -0.010639  -0.008708  -0.002372   

         Channel 7  Channel 8  Channel 9  Channel 10  ...  Channel 17  \
0        -0.007582  -0.013501  -0.004942   -0.015260  ...   -0.001552   
1        -0.007580  -0.013510  -0.004952   -0.015253  ...   -0.001557   
2        -0.007581  -0.013518  -0.004947   -0.015252  ...   -0.001553   
3        -0.007580  -0.013518  -0.004939   -0.015238  ...   -0.001557   
4        -0.007583  -0.013516  -0.004932   -0.015227  ...   -0.001561   
...            ...        ...        ...         ...  ...         ...   
1169292  -0.007806  -0.012967  -0.004794   -0.015056  ...   -0.001071   
1169293  -0.007803  -0.012963  -0.004788   -0.015053  ...   -0.001072   
1169294  -0.007802  -0.012959  -0.004779   -0.015045  ...   -0.001076   
1169295  -0.007805  -0.012958  -0.004780   -0.015044  ...   -0.001078   
1169296  -0.007803  -0.012953  -0.004777   -0.015042  ...   -0.001076   

         Channel 18  Channel 19  Channel 20  Channel 21  Channel 22  \
0          0.004929   -0.005479    0.013104   -0.001809   -0.005824   
1          0.004926   -0.005480    0.013102   -0.001811   -0.005836   
2          0.004923   -0.005484    0.013102   -0.001813   -0.005841   
3          0.004921   -0.005482    0.013101   -0.001811   -0.005840   
4          0.004924   -0.005484    0.013099   -0.001807   -0.005833   
...             ...         ...         ...         ...         ...   
1169292    0.005596   -0.004699    0.013496   -0.001021   -0.005271   
1169293    0.005597   -0.004700    0.013497   -0.001027   -0.005270   
1169294    0.005601   -0.004700    0.013501   -0.001032   -0.005267   
1169295    0.005603   -0.004699    0.013500   -0.001030   -0.005269   
1169296    0.005604   -0.004697    0.013498   -0.001031   -0.005270   

         Channel 23  Channel 24  Marker value  Sequence timestamp  
0         -0.001093    0.008498  Palm Forward                 614  
1         -0.001099    0.008493  Palm Forward                 615  
2         -0.001100    0.008490  Palm Forward                 616  
3         -0.001100    0.008489  Palm Forward                 617  
4         -0.001101    0.008488  Palm Forward                 618  
...             ...         ...           ...                 ...  
1169292   -0.000461    0.008163  Palm Forward             2223713  
1169293   -0.000462    0.008160  Palm Forward             2223714  
1169294   -0.000464    0.008158  Palm Forward             2223715  
1169295   -0.000466    0.008155  Palm Forward             2223716  
1169296   -0.000466    0.008152  Palm Forward             2223717  

[1169297 rows x 26 columns]

In [6]:
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.model_selection import train_test_split

df_X = df_sampled.drop(columns=['Marker value', 'Sequence timestamp'])  # Добавить 'Sequence timestamp' для выключения временных меток в train-сете, убрать для включения
df_y = df_sampled['Marker value']

le = LabelEncoder()
df_y = le.fit_transform(df_y)

X_train, X_test, y_train, y_test = train_test_split(df_X, df_y, test_size=0.3, random_state=42)

X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
y_train = y_train.astype('int32')
y_test = y_test.astype('int32')

sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

In [7]:
import _pickle as pkl
with open(f'scaler_v{data_version}.pkl', 'wb') as f:
    pkl.dump(sc, f)
with open(f'labels_v{data_version}.pkl', 'wb') as f:
    pkl.dump(le, f)

In [8]:
from keras.layers import Dense, Input, Dropout
from keras import Sequential

model = Sequential([
    Input(shape=(X_train.shape[1],)),
    Dense(128, activation='relu'),
    Dense(len(le.classes_), activation='softmax'),
])

model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 128)               3200      
                                                                 
 dense_1 (Dense)             (None, 4)                 516       
                                                                 
Total params: 3,716
Trainable params: 3,716
Non-trainable params: 0
_________________________________________________________________


In [9]:
model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=15)

Epoch 1/15
25579/25579 [==============================] - 78s 3ms/step - loss: 0.6730 - accuracy: 0.7331 - val_loss: 0.3852 - val_accuracy: 0.8766
Epoch 2/15
25579/25579 [==============================] - 80s 3ms/step - loss: 0.2943 - accuracy: 0.9020 - val_loss: 0.2061 - val_accuracy: 0.9378
Epoch 3/15
25579/25579 [==============================] - 101s 4ms/step - loss: 0.1865 - accuracy: 0.9397 - val_loss: 0.1540 - val_accuracy: 0.9495
Epoch 4/15
25579/25579 [==============================] - 78s 3ms/step - loss: 0.1384 - accuracy: 0.9555 - val_loss: 0.1206 - val_accuracy: 0.9624
Epoch 5/15
25579/25579 [==============================] - 77s 3ms/step - loss: 0.1115 - accuracy: 0.9639 - val_loss: 0.0945 - val_accuracy: 0.9699
Epoch 6/15
25579/25579 [==============================] - 79s 3ms/step - loss: 0.0947 - accuracy: 0.9694 - val_loss: 0.0923 - val_accuracy: 0.9703
Epoch 7/15
25579/25579 [==============================] - 100s 4ms/step - loss: 0.0833 - accuracy: 0.9729 - val_loss:

In [10]:
loss, accuracy = model.evaluate(X_test, y_test)

print(f'Loss: {loss:.4f}')
print(f'Accuracy: {accuracy:.4f}')

10963/10963 [==============================] - 20s 2ms/step - loss: 0.0346 - accuracy: 0.9893
Loss: 0.0346
Accuracy: 0.9893


### Experiment results
1. **Base exerimental** accuracy Train: `0.6697` @ `25`, Val: `0.7566` @ `25`
2. **Timestamped exerimental** accuracy Train: `0.6506` @ `10`, Val: `0.7557` @ `10`
3. **Base no-offset exerimental** accuracy Train: `0.6238` @ `10`, Val: `0.6786` @ `10`
4. **Timestamped no-offset exerimental** accuracy Train: `0.6246` @ `10`, Val: `0.6991` @ `10`
5. **Base no-dropout exerimental** accuracy Train: `0.9822` @ `10`, Val: `0.9807` @ `10` **Best**
6. **Base no-dropout data v2 exerimental** accuracy Train: `0.8568` @ `10`, Val: `0.8628` @ `10`

**Base no-dropout final** accuracy Train: `0.9837` @ `15`, Val: `0.9893` @ `15`


In [11]:
model.save(f'model_new_v{data_version}.keras')

In [12]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

y_pred = np.argmax(model.predict(X_test), axis=-1).astype('int32')
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')

print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))
print("Classification Report:")
print(classification_report(y_test, y_pred))

10963/10963 [==============================] - 9s 855us/step
Accuracy: 0.99
Confusion Matrix:
[[ 43894     15     16    277]
 [     3  42895     24    160]
 [    25     28  87588    860]
 [   251    683   1407 172664]]
Classification Report:
              precision    recall  f1-score   support

           0       0.99      0.99      0.99     44202
           1       0.98      1.00      0.99     43082
           2       0.98      0.99      0.99     88501
           3       0.99      0.99      0.99    175005

    accuracy                           0.99    350790
   macro avg       0.99      0.99      0.99    350790
weighted avg       0.99      0.99      0.99    350790



In [23]:
def test_sample(_idx):
    print(f'Sample #{_idx:06d}, true: {le.inverse_transform([y_test[_idx]])[0]}, pred: {le.inverse_transform([y_pred[_idx]])[0]}')

test_sample(0)
test_sample(2312)
test_sample(4331)
test_sample(23422)
test_sample(24212)
test_sample(124124)

Sample #000000, true: Flexion, pred: Flexion
Sample #002312, true: Palm Forward, pred: Palm Forward
Sample #004331, true: Pause, pred: Pause
Sample #023422, true: Palm Forward, pred: Palm Forward
Sample #024212, true: Palm Forward, pred: Palm Forward
Sample #124124, true: Pause, pred: Pause


In [26]:
np.savetxt('X_test.csv', X_test, delimiter=',')
np.savetxt('y_test.csv', y_test, delimiter=',', fmt='%d')

In [25]:
import os
import tensorflow.lite as tflite

converter = tflite.TFLiteConverter.from_keras_model(model)
tflite_model = converter.convert()

with open('model.tflite', 'wb') as f:
    f.write(tflite_model)

file_size = os.path.getsize("model.tflite")
print(f"Размер файла: {file_size} байт")

INFO:tensorflow:Assets written to: C:\Users\ZONT_\AppData\Local\Temp\tmpwu353_m8\assets
Размер файла: 16544 байт
